nion/instrumentation/DriftTracker.py

```diff
@@ -78,6 +81,114 @@ def __drift_changed(self, offset_nm: Geometry.FloatSize, elapsed_time: float) ->
         self.__event_loop.call_soon_threadsafe(functools.partial(self.__update_drift_log_data_item, delta_nm_data))
 
 
+@dataclasses.dataclass
+class DriftResult:
+    """Result of drift calculation returned by DriftCalculator.calculate
+
+    It can interpreted as: Between 'time_window_start' and 'time_window_end' the average drift rate was estimated to
+    be 'drift_rate'.
+
+    'drift_rate' should be in units of nm / s.
+    """
+
+    drift_rate: Geometry.FloatSize
+    time_window_start: datetime.datetime
+    time_window_end: datetime.datetime
+
+
+class DriftCalculator:
+    """Calculate drift rate from images with timestamps
+
+    Used by DriftTracker to perform the actual drift calculation from submitted images.
+
+    There can be different algorithms for drift calculation. Each of them should be implemented in its own class
+    which should be a subclass of this one.
+
+    DriftTracker discovers available drift calculation algorithms from the registry by looking for the component id
+    'drift_calculator'.
+
+    You need to set 'drift_calculator_id' to a unique name that will be used to identify a specifc algorithm.
+    'drift_calculator_name' is optional and will be used as the display name in a UI based drift calculator selection.
+    """
+    drift_calculator_id: str
+    drift_calculator_name: typing.Optional[str] = None
+
+    def calculate(self, first_xdata: typing.Optional[DataAndMetadata.DataAndMetadata], xdata_history: collections.deque) -> typing.Optional[DriftResult]:
+        ...
+
+
+class SimpleDriftCalculator(DriftCalculator):
+    drift_calculator_id = 'simple_drift_calculator'
+
+    def calculate(self, first_xdata: typing.Optional[DataAndMetadata.DataAndMetadata], xdata_history: collections.deque) -> typing.Optional[DriftResult]:
+        current_xdata = xdata_history[-1] if len(xdata_history) else None
+
+        if first_xdata and current_xdata:
+            quality, raw_offset = xd.register_template(first_xdata, current_xdata)
+            delta_time = (current_xdata.timestamp - first_xdata.timestamp).total_seconds()
+            assert delta_time > 0.0
+            offset = Geometry.FloatPoint.make(typing.cast(typing.Tuple[float, float], raw_offset))
+            delta_nm = Geometry.FloatSize(
+                    h=current_xdata.dimensional_calibrations[0].convert_to_calibrated_size(offset.y),
+                    w=current_xdata.dimensional_calibrations[1].convert_to_calibrated_size(offset.x))
+
+            return DriftResult(drift_rate=delta_nm / delta_time, time_window_start=first_xdata.timestamp, time_window_end=current_xdata.timestamp)
+
+        return None
+
+
+Registry.register_component(SimpleDriftCalculator(), {"drift_calculator"})
+
+
+class DriftDataSource:
+    """Collects images from a camera or scan device and provides them to DiftTracker for calculating drift.
```

Comment:
"DiftTracker" -> "DriftTracker"

Reply:
I don't understand that. What do you mean by that?

Reply:
Ah, spotted the typo now...